packages/core/contracts/trader/uniswap-broker/UniswapBroker.sol

```diff
@@ -23,57 +23,56 @@ contract UniswapBroker {
+     * and the 1st value is the the denominator of the true price.
+     * @param maxSpendTokens array of unit to represent the max to spend in the two tokens.
      * @param to recipient of the trade proceeds.
-     * @param to deadline to limit when the trade can execute. If the tx is mined after this timestamp then revert.
+     * @param deadline to limit when the trade can execute. If the tx is mined after this timestamp then revert.
      */
-
     function swapToPrice(
+        bool tradingAsEOA,
         address uniswapRouter,
         address uniswapFactory,
-        address tokenA,
-        address tokenB,
-        uint256 truePriceTokenA,
-        uint256 truePriceTokenB,
-        uint256 maxSpendTokenA,
-        uint256 maxSpendTokenB,
+        address[2] memory swappedTokens,
+        uint256[2] memory truePriceTokens,
+        uint256[2] memory maxSpendTokens,
         address to,
         uint256 deadline
     ) public {
         IUniswapV2Router01 router = IUniswapV2Router01(uniswapRouter);
 
         // true price is expressed as a ratio, so both values must be non-zero
-        require(truePriceTokenA != 0 && truePriceTokenB != 0, "SwapToPrice: ZERO_PRICE");
+        require(truePriceTokens[0] != 0 && truePriceTokens[1] != 0, "SwapToPrice: ZERO_PRICE");
         // caller can specify 0 for either if they wish to swap in only one direction, but not both
-        require(maxSpendTokenA != 0 || maxSpendTokenB != 0, "SwapToPrice: ZERO_SPEND");
+        require(maxSpendTokens[0] != 0 || maxSpendTokens[1] != 0, "SwapToPrice: ZERO_SPEND");
 
         bool aToB;
         uint256 amountIn;
         {
-            (uint256 reserveA, uint256 reserveB) = getReserves(uniswapFactory, tokenA, tokenB);
-            (aToB, amountIn) = computeTradeToMoveMarket(truePriceTokenA, truePriceTokenB, reserveA, reserveB);
+            (uint256 reserveA, uint256 reserveB) = getReserves(uniswapFactory, swappedTokens[0], swappedTokens[1]);
+            (aToB, amountIn) = computeTradeToMoveMarket(truePriceTokens[0], truePriceTokens[1], reserveA, reserveB);
         }
 
         require(amountIn > 0, "SwapToPrice: ZERO_AMOUNT_IN");
 
         // spend up to the allowance of the token in
-        uint256 maxSpend = aToB ? maxSpendTokenA : maxSpendTokenB;
+        uint256 maxSpend = aToB ? maxSpendTokens[0] : maxSpendTokens[1];
         if (amountIn > maxSpend) {
             amountIn = maxSpend;
         }
 
-        address tokenIn = aToB ? tokenA : tokenB;
-        address tokenOut = aToB ? tokenB : tokenA;
-        TransferHelper.safeTransferFrom(tokenIn, msg.sender, address(this), amountIn);
+        address tokenIn = aToB ? swappedTokens[0] : swappedTokens[1];
+        address tokenOut = aToB ? swappedTokens[1] : swappedTokens[0];
+
         TransferHelper.safeApprove(tokenIn, address(router), amountIn);
 
+        if (tradingAsEOA) TransferHelper.safeTransferFrom(tokenIn, msg.sender, address(this), amountIn);
```

Comment:
that's correct. Only if it's trading from an EOA do we need to move the tokens into this contract. if it's called from a DSProxy then the context of `this` becomes the DSProxy and the logic within the `UniswapBroker` is effectively executed _as_ the DSProxy.

Reply:
Is there a similar transfer _out_ that needs to be enabled/disabled?

Reply:
right now the contract should not be left with any extra tokens are a trade; it pulls the exact amount needed for the trade and sends everything left over to the calling EOA.

This does open up a separate meta point that I think we can leave for a bit later: we'll want some more general DSProxy methods for doing things like:
- deposit/withdraw/ transfer tokens from the DSProxy
- wrapper methods for dealing with EMPs/perps
- admin methods for adding other admins etc

Reply:
the `to` parameter ensures this. If an EOA is calling this then they can set the `to` to be their address. This will send the output of the trade to the EOA. In the DSProxy case the caller must set the `to` to be the DSProxy wallet. If this is set then the DSProxy will hold the tokens after the trade.

I've updated the comments in the `UniswapTrader` to reflect this fully.

Reply:
Instead of the `tradingAsEOA` step, why not just allow this contract to be withdrawn from? The downside is that it would have to be Ownable however, or store balances for account

Reply:
@chrismaree that makes sense! Thanks for the explanation!